src/method/method_runtime_context.cpp

```diff
@@ -0,0 +1,229 @@
+    return rctx;
+  }
+
+//////////////////////////////////////////////////////////////////////////
+// Runtime Context
+//////////////////////////////////////////////////////////////////////////
+
+  runtime_context::runtime_context ()
+    : m_group_stack {}
+    , m_returning_cursors {}
+    , m_group_map {}
+    , m_cursor_map {}
+  {
+    //
+  }
+
+  runtime_context::~runtime_context ()
+  {
+    destroy_all_cursors ();
+    destroy_all_groups ();
+  }
+
+  method_invoke_group *
+  runtime_context::create_invoke_group (cubthread::entry *thread_p, const method_sig_list &sig_list)
+  {
+    method_invoke_group *group = new (std::nothrow) cubmethod::method_invoke_group (thread_p, sig_list);
+    if (group)
+      {
+	m_group_map [group->get_id ()] = group;
+	// TODO
+      }
+    return group;
+  }
+
+  void
+  runtime_context::push_stack (method_invoke_group *group)
+  {
+    m_group_stack.push_back (group->get_id ());
+  }
+
+  void
+  runtime_context::pop_stack ()
+  {
+    m_group_stack.pop_back ();
+
+    if (m_group_stack.empty ())
+      {
+	// TODO
+      }
+  }
+
+  method_invoke_group *
+  runtime_context::top_stack ()
+  {
+    assert (m_group_stack.empty () == false);
+
+    METHOD_GROUP_ID top = m_group_stack.back ();
+
+    const auto &it = m_group_map.find (top);
+    if (it == m_group_map.end ())
+      {
+	// should not happended
+	assert (false);
+	return nullptr;
+      }
+
+    return it->second;;
+  }
+
+  query_cursor *
+  runtime_context::get_cursor (cubthread::entry *thread_p, QUERY_ID query_id)
+  {
+    if (query_id == NULL_QUERY_ID)
+      {
+	return nullptr;
+      }
+
+    // find in map
+    auto search = m_cursor_map.find (query_id);
+    if (search != m_cursor_map.end ())
+      {
+	// found
+	return search->second;
+      }
+
+    return nullptr;
+  }
+
+  query_cursor *
+  runtime_context::create_cursor (cubthread::entry *thread_p, QUERY_ID query_id, bool is_oid_included)
+  {
+    if (query_id == NULL_QUERY_ID || query_id >= SHRT_MAX)
+      {
+	// something wrong!
+	assert (false);
+	return nullptr;
+      }
+
+    query_cursor *cursor = nullptr;
+
+    // find in map
+    auto search = m_cursor_map.find (query_id);
+    if (search != m_cursor_map.end ())
+      {
+	// found
+	cursor = search->second;
+	if (cursor)
+	  {
+	    cursor->change_owner (thread_p);
+	    return cursor;
+	  }
+      }
+    else
+      {
+	// not found, create a new server-side cursor
+	int tran_index = LOG_FIND_THREAD_TRAN_INDEX (thread_p);
+	QMGR_QUERY_ENTRY *query_entry_p = qmgr_get_query_entry (thread_p, query_id, tran_index);
+	if (query_entry_p != NULL)
+	  {
+	    // m_list_id is going to be destoryed on server-side, so that qlist_count has to be updated
+	    qfile_update_qlist_count (thread_p, query_entry_p->list_id, 1);
+
+	    // store a new cursor in map
+	    cursor = new (std::nothrow) query_cursor (thread_p, query_entry_p, is_oid_included);
+	    m_cursor_map [query_id] = cursor;
+	  }
+      }
+
```

Comment:
Wouldn't the assert(cursor != nullptr) check be necessary?

Reply:
I've added it. thank you.